charmcraft/package.py

```diff
@@ -418,3 +423,123 @@ def handle_package(self, prime_dir, bases_config: BasesConfiguration):
 
         zipfh.close()
         return zipname
+
+    def _get_charm_pack_args(self, base_indeces: List[str], destructive_mode: bool) -> List[str]:
+        """Get the arguments for a charmcraft pack subprocess to run."""
+        args = ["charmcraft", "pack", "--verbose"]
+        if destructive_mode:
+            args.append("--destructive-mode")
+        for base in base_indeces:
+            args.append(f"--bases-index={base}")
+        if self.force_packing:
+            args.append("--force")
+        return args
+
+    def pack_bundle(
+        self,
+        *,
+        charms: Dict[str, pathlib.Path],
+        base_indeces: List[str],
+        destructive_mode: bool,
+        overwrite: bool = False,
+    ) -> List[pathlib.Path]:
+        """Pack a bundle."""
+        if self._parts is None:
+            self._parts = {"bundle": {"plugin": "bundle"}}
+
+        if env.is_charmcraft_running_in_managed_mode():
+            work_dir = env.get_managed_environment_home_path()
+        else:
+            work_dir = self.config.project.dirpath / const.BUILD_DIRNAME
+
+        # get the config files
+        bundle_filepath = self.config.project.dirpath / "bundle.yaml"
+        bundle = load_yaml(bundle_filepath)
+        bundle_name = bundle.get("name")
+        if not bundle_name:
+            raise CraftError(
+                "Invalid bundle config; missing a 'name' field indicating the bundle's name in "
+                "file {!r}.".format(str(bundle_filepath))
+            )
+
+        if charms:
+            bundle_charms = bundle.get("applications", {})
+            command_args = self._get_charm_pack_args(base_indeces, destructive_mode)
+            charms = _subprocess_pack_charms(charms, command_args)
+            for name, value in bundle_charms.items():
+                if name in charms:
+                    value["charm"] = charms[name]
+        else:
+            charms = {}
+
+        lifecycle = parts.PartsLifecycle(
+            self._parts,
+            work_dir=work_dir,
+            project_dir=self.config.project.dirpath,
+            project_name=bundle_name,
+            ignore_local_sources=[bundle_name + ".zip"],
+        )
+
+        lifecycle.run(parts.Step.PRIME)
+
+        # pack everything
+        create_manifest(
+            lifecycle.prime_dir,
+            self.config.project.started_at,
+            bases_config=None,
+            linting_results=[],
+        )
+        zipname = self.config.project.dirpath / (bundle_name + ".zip")
+        if overwrite:
+            primed_bundle_path = lifecycle.prime_dir / "bundle.yaml"
+            with primed_bundle_path.open("w") as bundle_file:
+                yaml.safe_dump(bundle, bundle_file)
+        build_zip(zipname, lifecycle.prime_dir)
+
+        return [*charms.values(), zipname]
```

Comment:
What do you think if we `return zipname, [*charms.values(), zipname]`? Get a [-1] as filename feels unintuitive.

Reply:
What about returning (as pseudocode) `(charm_paths, [bundle_path])`? (Probably as a data structure)

Then in the outer command we can add the list of generated charms to the JSON?

Reply:
We don't seems use the `output_files` except the zipname? `_pack_bundle` return the list but nobody use it?

Reply:
I've changed this around a bit — you and @cmatsuoka tell me what you think

Reply:
LGTM